File: wallet_scanner/scanner/output.py
import csv
import json
import logging
import os
from datetime import datetime, timezone
from typing import Optional

logger = logging.getLogger(__name__)


def get_timestamp() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def format_filename() -> str:
    return datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")


def aggregate_results(
    all_addresses: dict[str, list[dict]],
    mnemonic_index: int,
    snapshot_time: str,
) -> dict:
    addresses: list[dict] = []
    total_usd_value = 0.0

    for chain, results in all_addresses.items():
        for addr_result in results:
            address = addr_result.get("address", "")

            if chain.lower() == "solana":
                addr_entry = {
                    "chain": chain.lower(),
                    "address": address,
                    "sol_balance": addr_result.get("sol", 0.0),
                    "usdt": addr_result.get("usdt", 0.0),
                    "usdc": addr_result.get("usdc", 0.0),
                    "total_usd_value": 0.0,
                }
            else:
                native_balance = addr_result.get("native_balance", 0.0) or 0.0
                addr_entry = {
                    "chain": chain.lower(),
                    "address": address,
                    "native_balance": native_balance,
                    "usdt": addr_result.get("usdt", 0.0),
                    "usdc": addr_result.get("usdc", 0.0),
                    "total_usd_value": 0.0,
                }

            addresses.append(addr_entry)

    return {
        "mnemonic_index": mnemonic_index,
        "addresses": addresses,
        "total_usd_value": total_usd_value,
        "snapshot_time": snapshot_time,
    }


def write_json_output(
    output_dir: str,
    results: list[dict],
    threshold_usd: float,
    scan_time: Optional[str] = None,
) -> str:
    if scan_time is None:
        scan_time = get_timestamp()

    os.makedirs(output_dir, exist_ok=True)

    timestamp_str = format_filename()
    filename = f"{timestamp_str}_scan_results.json"
    filepath = os.path.join(output_dir, filename)

    total_scanned = len(results)
    passed = sum(1 for r in results if r.get("total_usd_value", 0.0) >= threshold_usd)

    output = {
        "scan_time": scan_time,
        "threshold_usd": threshold_usd,
        "total_scanned": total_scanned,
        "passed": passed,
        "results": results,
    }

    try:
        with open(filepath, "w", encoding="utf-8") as f:
            json.dump(output, f, indent=2, ensure_ascii=False)
        logger.info(f"JSON output written to {filepath}")
    except Exception as e:
        logger.error(f"Failed to write JSON output: {e}")
        raise

    return filepath


def write_csv_output(
    output_dir: str,
    results: list[dict],
    scan_time: Optional[str] = None,
) -> str:
    if scan_time is None:
        scan_time = get_timestamp()

    os.makedirs(output_dir, exist_ok=True)

    timestamp_str = format_filename()
    filename = f"{timestamp_str}_scan_results.csv"
    filepath = os.path.join(output_dir, filename)

    rows = []
    for result in results:
        mnemonic_index = result.get("mnemonic_index")
        snapshot_time = result.get("snapshot_time", scan_time)
        addresses = result.get("addresses", [])

        for addr in addresses:
            chain = addr.get("chain", "")
            native_bal = addr.get("native_balance") if chain != "solana" else addr.get("sol_balance", 0.0)
            row = {
                "mnemonic_index": mnemonic_index,
                "chain": chain,
                "address": addr.get("address", ""),
                "native_balance": native_bal,
                "usdt_balance": addr.get("usdt", 0.0),
                "usdc_balance": addr.get("usdc", 0.0),
                "total_usd_value": addr.get("total_usd_value", 0.0),
                "snapshot_time": snapshot_time,
            }
            rows.append(row)

    try:
        with open(filepath, "w", encoding="utf-8", newline="") as f:
            if rows:
                writer = csv.DictWriter(
                    f,
                    fieldnames=[
                        "mnemonic_index",
                        "chain",
                        "address",
                        "native_balance",
                        "usdt_balance",
                        "usdc_balance",
                        "total_usd_value",
                        "snapshot_time",
                    ],
                )
                writer.writeheader()
                writer.writerows(rows)
        logger.info(f"CSV output written to {filepath}")
    except Exception as e:
        logger.error(f"Failed to write CSV output: {e}")
        raise

    return filepath

File: wallet_scanner/src/output.ts
import fs from "node:fs";
import path from "node:path";
import { stringify } from "csv-stringify/sync";
import type { AggregatedResult, Chain, EvmAddressBalance, SolanaAddressBalance } from "./types.js";

export const getTimestamp = (): string => new Date().toISOString().replace(/\.\d{3}Z$/, "Z");

export const formatFilename = (): string => {
  const d = new Date();
  const yyyy = d.getUTCFullYear().toString().padStart(4, "0");
  const mm = (d.getUTCMonth() + 1).toString().padStart(2, "0");
  const dd = d.getUTCDate().toString().padStart(2, "0");
  const hh = d.getUTCHours().toString().padStart(2, "0");
  const mi = d.getUTCMinutes().toString().padStart(2, "0");
  const ss = d.getUTCSeconds().toString().padStart(2, "0");
  return `${yyyy}${mm}${dd}_${hh}${mi}${ss}`;
};

export const aggregateResults = (
  allAddresses: Record<string, Array<EvmAddressBalance | SolanaAddressBalance>>,
  mnemonicIndex: number,
  snapshotTime: string
): AggregatedResult => {
  const addresses: AggregatedResult["addresses"] = [];

  for (const [chain, results] of Object.entries(allAddresses)) {
    for (const addrResult of results) {
      const address = (addrResult as any).address ?? "";
      const chainLower = chain.toLowerCase() as Chain;

      if (chainLower === "solana") {
        addresses.push({
          chain: chainLower,
          address,
          sol_balance: (addrResult as any).sol ?? 0.0,
          usdt: (addrResult as any).usdt ?? 0.0,
          usdc: (addrResult as any).usdc ?? 0.0,
          total_usd_value: 0.0
        });
      } else {
        addresses.push({
          chain: chainLower,
          address,
          native_balance: (addrResult as any).native_balance ?? 0.0,
          usdt: (addrResult as any).usdt ?? 0.0,
          usdc: (addrResult as any).usdc ?? 0.0,
          total_usd_value: 0.0
        });
      }
    }
  }

  return {
    mnemonic_index: mnemonicIndex,
    addresses,
    total_usd_value: 0.0,
    snapshot_time: snapshotTime
  };
};

export const writeJsonOutput = (
  outputDir: string,
  results: AggregatedResult[],
  thresholdUsd: number,
  scanTime?: string
): string => {
  const effectiveScanTime = scanTime ?? getTimestamp();
  fs.mkdirSync(outputDir, { recursive: true });

  const filename = `${formatFilename()}_scan_results.json`;
  const filepath = path.join(outputDir, filename);

  const totalScanned = results.length;
  const passed = results.filter((r) => (r.total_usd_value ?? 0.0) >= thresholdUsd).length;

  const output = {
    scan_time: effectiveScanTime,
    threshold_usd: thresholdUsd,
    total_scanned: totalScanned,
    passed,
    results
  };

  fs.writeFileSync(filepath, JSON.stringify(output, null, 2), "utf-8");
  return filepath;
};

export const writeCsvOutput = (
  outputDir: string,
  results: AggregatedResult[],
  scanTime?: string
): string => {
  const effectiveScanTime = scanTime ?? getTimestamp();
  fs.mkdirSync(outputDir, { recursive: true });

  const filename = `${formatFilename()}_scan_results.csv`;
  const filepath = path.join(outputDir, filename);

  const rows: Array<Record<string, string | number>> = [];
  for (const result of results) {
    const mnemonicIndex = result.mnemonic_index;
    const snapshotTime = result.snapshot_time || effectiveScanTime;

    for (const addr of result.addresses) {
      const nativeBal = addr.chain === "solana" ? (addr.sol_balance ?? 0.0) : (addr.native_balance ?? 0.0);
      rows.push({
        mnemonic_index: mnemonicIndex,
        chain: addr.chain,
        address: addr.address,
        native_balance: nativeBal,
        usdt_balance: addr.usdt ?? 0.0,
        usdc_balance: addr.usdc ?? 0.0,
        total_usd_value: addr.total_usd_value ?? 0.0,
        snapshot_time: snapshotTime
      });
    }
  }

  const csv = stringify(rows, {
    header: true,
    columns: [
      "mnemonic_index",
      "chain",
      "address",
      "native_balance",
      "usdt_balance",
      "usdc_balance",
      "total_usd_value",
      "snapshot_time"
    ]
  });

  fs.writeFileSync(filepath, csv, "utf-8");
  return filepath;
};

